allocations/allocation_mapper.py

```diff
@@ -11,45 +34,129 @@ def _wants_db_allocations(self, request):
                     if 'used' in active_allocation and active_allocation['used']:
-                        alloc.su_used = float(balance_service.call(project.chargeCode)['used'])
-                reformated_allocations.append(self.portal_to_tas_obj(alloc))
-            project.allocations = reformated_allocations
+                        alloc.su_used = float(balance_service.call(project_charge_code)['used'])
+                reformated_alloc = self.portal_to_tas_obj(alloc)
+                if is_dict_type:
+                    reformated_alloc = reformated_alloc.__dict__
+                reformated_allocations.append(reformated_alloc)
+            if is_dict_type:
+                project['allocations'] = reformated_allocations
+            else:
+                project.allocations = reformated_allocations
         return project
+    
+    def save_allocation(self, alloc, project_charge_code, tas):
+        if self.allocations_from_db:
+            reformated_alloc = self.tas_to_portal_obj(alloc, project_charge_code)
+            reformated_alloc.save()
+            # notify Chameleon operators
+            subject = 'Pending allocation request for project {}'.format(project_charge_code)
+            body = '''
+                    <p>Please review the pending allocation request for project {project_charge_code}
+                    at <a href='https://www.chameleoncloud.org/admin/allocations/' target='_blank'>admin page</a></p>
+                    '''.format(project_charge_code = project_charge_code)
+            send_mail(subject, strip_tags(body), settings.DEFAULT_FROM_EMAIL, [settings.PENDING_ALLOCATION_NOTIFICATION_EMAIL], html_message=body)
+        else:
+            tas.create_allocation(alloc)
+
+    def get_user_id(self, request, tas):
+        if self.allocations_from_db:
+            return request.user.id
+        else:
+            return tas.get_user(username=request.user)['id']
+
+    def get_username(self, request, tas):
+        if self.allocations_from_db:
+            return request.user.username
+        else:
+            return tas.get_user(username=request.user)['username']
+
+    def allocation_approval(self, data, tas):
+        if self.allocations_from_db:
+            # update allocation model
+            alloc = portal_alloc.objects.filter(id=data['id'])[0]
+            data['status'] = data['status'].lower()
+            for item in ['reviewerId', 'dateReviewed', 'start', 'end', 'status', 'decisionSummary', 'computeAllocated']:
+                setattr(alloc, TAS_TO_PORTAL_MAP[item], data[item])
+            alloc.save()
+            logger.info('Allocation model updated: data=%s', alloc.__dict__)
+            # send email to PI
+            user_model = get_user_model()
+            user = user_model.objects.filter(id=data['requestorId'])[0]
+            subject = 'Decision of your allocation request for project {}'.format(data['project'])
```

Comment:
Would also clean up this function if the email sending was a private helper function